### ISSUE TITLE
Introduce gradient-checkpointing

### ISSUE THREAD

**Issue by dominikandreasseitz** [pasqal-io/pyqtorch#62]
Use the gradient checkpointing technique to save memory, especially for higher-order derivatives. PyTorch already offers routines for implementing checkpointing, as detailed here: https://pytorch.org/docs/stable/checkpoint.html.

Tasks to accomplish:

- [ ] build a suitable benchmark using a loss function that includes higher-order derivatives
- [ ] implement the gradient checkpointing for generic models
- [ ] test gains in memory consumption for large batch sizes

**Comment by dominikandreasseitz** [comment]
can be closed since adjoint differentiation has been added
